sklearn/metrics/pairwise.py

```diff
@@ -935,13 +939,48 @@ def _parallel_pairwise(X, Y, func, n_jobs, **kwds):
     if Y is None:
         Y = X
 
+    if n_jobs == 1:
+        # Special case to avoid picklability checks in delayed
+        return func(X, Y, **kwds)
+
+    # TODO: in some cases, backend='threading' may be appropriate
+    fd = delayed(func)
     ret = Parallel(n_jobs=n_jobs, verbose=0)(
-        delayed(func)(X, Y[s], **kwds)
+        fd(X, Y[s], **kwds)
         for s in gen_even_slices(Y.shape[0], n_jobs))
 
     return np.hstack(ret)
 
 
+def _pairwise_callable(X, Y, metric, **kwds):
+    """Handle the callable case for pairwise_{distances,kernels}
+    """
+    X, Y = check_pairwise_arrays(X, Y)
+    if X is Y:
+        # Only calculate metric for upper triangle
+        out = np.zeros((X.shape[0], Y.shape[0]), dtype='float')
+        iterator = itertools.combinations(range(X.shape[0]), 2)
+        iterator = itertools.combinations(range(X.shape[0]), 2)
+    else:
+        # Calculate all cells
+        out = np.empty((X.shape[0], Y.shape[0]), dtype='float')
+        iterator = itertools.product(range(X.shape[0]), range(Y.shape[0]))
+
+    for i, j in iterator:
+        out[i, j] = metric(X[i], Y[j], **kwds)
```

Comment:
Yes, I think you're right. The joint work made more sense before I realised it was necessary to fill the diagonal in all cases...